image-processing.js

```diff
@@ -0,0 +1,59 @@
+'use strict'
+const R = require('ramda')
+const stream = require('stream')
+const sharp = require('sharp')
+
+const renameThumb = (name, options) => {
+  const dotIndex = name.lastIndexOf('.')
+  const dotExtension = name.slice(dotIndex)
+  const filename = name.slice(0, dotIndex)
+  const label = options.label || `@${options.width}x${options.height}`
+
+  return `${filename}${label}${dotExtension}`
+}
+
+module.exports = (client, inputArgs, cache) => {
+  const name = inputArgs.name
+  const data = inputArgs.data
+  const options = inputArgs.options
+  let originalData, fullsize
+
+  if ((inputArgs.data instanceof stream.Stream)) {
+    originalData = new stream.PassThrough()
+    data.pipe(originalData)
+    fullsize = originalData.pipe(sharp().resize(options.maxSize, options.maxSize).max().withoutEnlargement().on('info', info => options.meta = info))
+  } else {
+    originalData = data
+    fullsize = sharp(originalData).resize(options.maxSize, options.maxSize).max().withoutEnlargement().on('info', info => options.meta = info)
+  }
+
+  const originalUpload = client.upload(name, fullsize, options).then(cache.put(name, data))
+  const uploads = [originalUpload]
+
+  // add check for mimetype before trying to do image processing
+
+  if (options.thumbnails && Array.isArray(options.thumbnails)) {
+    options.thumbnails.forEach(thumbOptions => {
+      const thumbname = renameThumb(name, thumbOptions)
+      const addMeta = meta => versionOptions.meta = meta
+      const versionOptions = R.clone(options)
+      versionOptions.isThumb = thumbOptions.isThumb
+      let resizedData
+
+      if (data instanceof stream.Stream) {
+        const dataCopy = new stream.PassThrough()
```

Comment:
It's not clear from the Node stream docs what this does here.  
> The stream.PassThrough class is a trivial implementation of a Transform stream that simply passes the input bytes across to the output.
Are you creating a stream out of a non-stream?